### PR TITLE
Fix probable typo in value of `MAX_TILE_ELEMENTS`

### PR DESCRIPTION
**Pull request by Broxzier**
`MAX_TILE_ELEMENTS` was defined as 196096, which is 0x2FE00, but the comment behind it says 0x30000.

I'm not sure the difference of 512 is there for a reason or that it was a typo. I'd be nice if someone could verify this.

### CODE REVIEW

**Comment by Gymnasiast** [comment]
RCT2_MAX_TILE_ELEMENTS is 0x30000, so most likely just a typo.

**Comment by janisozaur** [comment]
Many people on NE hit those limits, we could advertise it as we have updated some of them already, even if just by a small value, 2.6‰ :stuck_out_tongue:

**Comment by Broxzier** [comment]
This crashes when going over the old limit, apparently reordering the tile elements fails somehow.

**Comment by IntelOrca** [comment]
Although `MAX_ELEMENTS` had the incorrect value the number of elements actually stored in memory was still correct as it used `MAX_TILE_TILE_ELEMENT_POINTERS * 3` (don't ask me why).

@Broxzier I have refactored some of the map reorganisation code, so could you try it again and see if it still crashes?

In particular I have removed `sub_68B089`, this function was run every game tick and 1000 times when trying to request some spare elements. It was a lighter function which tried to stretch out the tile element pointers behind any spare elements. In this day and age, doing a full tile element defrag is very fast (even in debug mode it takes no more than 10ms), so I don't think keeping sub_68B089 is necessary.

**Comment by Broxzier** [comment]
@IntelOrca Thanks for the refactor, that makes it much easier to understand. Sadly, this doesn't fix the issue. I think it doesn't have much to do with the reorganizing, but with how an elements are overwritten while inserting a new one. I've noticed that the first few elements in the list become invalid, but I haven't yet found out where exactly. I'll do some more debugging later tonight.

Here's a saved game in which I've been able to reproduce the crash consistently. Just place a few scenery objects (~20) after loading.
[Full Map - Riverland Season 9.zip](https://github.com/OpenRCT2/OpenRCT2/files/2116367/Full.Map.-.Riverland.Season.9.zip)

File: src/openrct2/world/Map.cpp [hunk 1]
```diff
@@ -3185,25 +3136,30 @@ void map_reorganise_elements()
...
-        return true;
-
-    for (sint32 i = 1000; i != 0; --i)
-        sub_68B089();
-
-    if ((gNextFreeTileElement + num_elements) <= gTileElements + MAX_TILE_ELEMENTS)
-        return true;
+    if (numElements != 0)
+    {
+        auto tileElementEnd = &gTileElements[MAX_TILE_ELEMENTS];
```
Comment (Broxzier) [inline src/openrct2/world/Map.cpp]: I think the index here needs to be `MAX_TILE_ELEMENTS - 1`, or the check below should be `>=` instead.

**Comment by IntelOrca** [comment]
Okay, I think the issue is that insert tile expects there to be enough spare elements for all the elements on the tile it is editing plus the extra element it is inserting. Perhaps the original maximum - 512 was a lazy way of ensuring that. In any case I think `tile_element_insert` needs to call `map_check_free_elements_and_reorganise` with the the number of elements on the tile + 1 rather than just 1.

**Comment by tupaschoal** [comment]
@IntelOrca and @Broxzier , I solved the merge conflict, is this mergeable as is or does it still need improvements?

**Comment by Gymnasiast** [comment]
@tupaschoal There is talk of the game crashing if we go over the old limit. Try loading a park that hits the tile element limit and see if you can remove and add tile elements without crashing.

An example of such a park is "Emerald Pointe" by "Coasterbill", which you can find on New Element:

[tile-element-limit-hit 191017-rct2-coasterbill-emerald_pointe.zip](https://github.com/OpenRCT2/OpenRCT2/files/4366456/tile-element-limit-hit.191017-rct2-coasterbill-emerald_pointe.zip)

**Comment by tupaschoal** [comment]
I couldn't reproduce the crash. I downloaded the park you mentioned, deleted a random track and tried to rebuild.

Both builds with and without the change can't rebuild the track just deleted, the message in game appears as "Can't construct this here (undefined string)", but the "develop" version shows "Not enough free map elements to place track" on the "TrackPlaceAction.hpp"

**Comment by Gymnasiast** [comment]
I have tested it with Emerald Pointe, by simply placing base blocks on top of each other. This crashes the game because of a null pointer getting hit in `map_reorganise_elements()`.

**Comment by Gymnasiast** [comment]
Investigation shows that there are more places where it uses the wrong value of `3 * MAX_TILE_ELEMENT_POINTERS`.

**Comment by duncanspumpkin** [comment]
Replaced by https://github.com/OpenRCT2/OpenRCT2/pull/11645

**Pull request closed without merging**